src/Riok.Mapperly/MapperGenerator.cs

```diff
@@ -22,39 +22,66 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
         var mapperClassDeclarations = SyntaxProvider.GetClassDeclarations(context);
 
         var compilationAndMappers = context.CompilationProvider.Combine(mapperClassDeclarations.Collect());
-        context.RegisterImplementationSourceOutput(compilationAndMappers, static (spc, source) => Execute(source.Left, source.Right, spc));
+        var mappersWithDiagnostics = compilationAndMappers.Select(
+            static (x, cancellationToken) => BuildDescriptors(x.Left, x.Right, cancellationToken)
+        );
+
+        // output the diagnostics
+        context.ReportDiagnostics(mappersWithDiagnostics.Select(static (source, _) => source.Diagnostics).WithTrackingName("diagnostics"));
```

Comment:
Since `WithTrackingName` is not available in older Roslyn versions we could probably just add an extension method `IncrementalValueProvider<TSource> WithTrackingName<TSource>(this IncrementalValueProvider<TSource> source, string name) => this;` to mock this for older roslyn versions.